app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -8,72 +11,98 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
-          set_flash_message(:alert, :failure, kind: OmniAuth::Utils.camelize(provider),
-            reason: @user.errors.full_messages.join(', '))
-        end
-        # Removing extra as it can overflow some session stores
-        session["devise.#{provider}_data"] = request.env['omniauth.auth'].except('extra')
-        redirect_to new_user_registration_url
+        register_new_user
       end
     end
 
     alias bird sso_callback
     alias nbp sso_callback
     alias mocksaml sso_callback
 
-    def provider
-      request.env['omniauth.auth']&.provider || params[:provider]
-    end
+    def deauthorize # rubocop:disable Metrics/AbcSize
+      identity = current_user.identities.find_by(omniauth_provider:)
+
+      if identity.nil?
+        set_flash(:alert, 'users.omni_auth.failure_deauthorize_not_linked')
+
+      elsif current_user.omniauth_identities.one? && !current_user.password_set?
+        set_flash(:alert, 'users.omni_auth.failure_deauthorize_last_identity')
 
-    def deauthorize # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
-      identity = current_user.identities.find_by(omniauth_provider: provider)
-      if is_navigational_format? && identity.nil?
-        # i18n-tasks-use t('users.omni_auth.failure_deauthorize_not_linked')
-        set_flash_message(:alert, :failure_deauthorize_not_linked, kind: OmniAuth::Utils.camelize(provider),
-          scope: 'users.omni_auth')
-      elsif is_navigational_format? && current_user.identities.count == 1 && !current_user.password_set?
-        # i18n-tasks-use t('users.omni_auth.failure_deauthorize_last_identity')
-        set_flash_message(:alert, :failure_deauthorize_last_identity, kind: OmniAuth::Utils.camelize(provider),
-          scope: 'users.omni_auth')
-      elsif is_navigational_format? && identity.destroy
-        remove_provider_from_session(provider)
-        # i18n-tasks-use t('users.omni_auth.success_deauthorize')
-        set_flash_message(:notice, :success_deauthorize, kind: OmniAuth::Utils.camelize(provider),
-          scope: 'users.omni_auth')
-      elsif is_navigational_format? && identity.errors.any?
-        # i18n-tasks-use t('users.omni_auth.failure_deauthorize')
-        set_flash_message(:alert, :failure_deauthorize, kind: OmniAuth::Utils.camelize(provider), scope: 'users.omni_auth',
-          reason: identity.errors.full_messages.join(', '))
+      elsif identity.destroy
+        remove_provider_from_session(omniauth_provider)
+        set_flash(:notice, 'users.omni_auth.success_deauthorize')
+
+      elsif identity.errors.any?
+        set_flash(:alert, 'users.omni_auth.failure_deauthorize', reason: identity.errors.full_messages.join(', '))
       end
       redirect_to edit_user_registration_path
     end
 
     private
 
+    def add_idp_to_existing_user # rubocop:disable Metrics/AbcSize
+      @current_user.identities << @user_identity
+      if @current_user.errors.blank?
```

Comment:
The line `@current_user.identities << @user_identity` aims to persist the new `@user_identity` and associates it with the `@current_user` (in a single transaction) [[Source]](https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many):

> **collection<<(object, ...)**
> Adds one or more objects to the collection by setting their foreign keys to the collection’s primary key. Note that this operation instantly fires update SQL without waiting for the save or update call on the parent object, unless the parent object is a new record. This will also run validations and callbacks of associated object(s).

As written in the documentation, it will validate the object, populating the `@current_user.errors` object if there are any (without saving the changes, of course). Thus, we use this information to handle invalid data.

Reply:
I tested it and it seems like the documentation is wrong? When adding an invalid user identity there are no validation errors (but it also does not get saved so apparently the validations do run?). While fixing this issue I also found another one: When adding an IdP that is already added to another account you just get signed in with that other account and there is no error message. That is weird. I will add a check that shows an error if the IdP account is already "used".

Reply:
You're absolutely right! The validation error is not present as expected. While I am generally not a big fan of saving something twice (through `<<` and `save`), the only improvement I could currently think of would be to call `u.valid?` rather than `u.save`.

<hr>

The other issue is, indeed, valid -- thanks for finding and fixing. I would be grateful if you could add a test for these two errors (validation and adding a previously-used IdP) :+1:

Reply:
Done.